include/execution.hpp

```diff
@@ -2461,7 +2461,110 @@ namespace _P2300::execution {
   /////////////////////////////////////////////////////////////////////////////
   // [execution.senders.adaptors.bulk]
   namespace __bulk {
+    template <class _ReceiverId, integral _Shape, class _FunId>
+      class __receiver
+        : receiver_adaptor<__receiver<_ReceiverId, _Shape, _FunId>, __t<_ReceiverId>> {
+        using _Receiver = __t<_ReceiverId>;
+        using _Fun = __t<_FunId>;
+        friend receiver_adaptor<__receiver, _Receiver>;
+
+        [[no_unique_address]] _Shape __shape_;
+        [[no_unique_address]] _Fun __f_;
+
+        template <class... _As>
+        void set_value(_As&&... __as) && noexcept 
+          requires __nothrow_invocable<_Fun, _Shape, _As...> {
+          for (_Shape __i{}; __i != __shape_; ++__i) {
+            std::invoke(__f_, __i, std::forward<_As>(__as)...);
+          }
+          execution::set_value(std::move(this->base()), std::forward<_As&&>(__as)...);
+        }
+
+        template <class... _As>
+        void set_value(_As&&... __as) && noexcept {
+          try {
+            for (_Shape __i{}; __i != __shape_; ++__i) {
+              std::invoke(__f_, __i, std::forward<_As>(__as)...);
+            }
+            execution::set_value(std::move(this->base()), std::forward<_As&&>(__as)...);
+          } catch(...) {
+            set_error(std::move(this->base()), std::current_exception());
+            execution::set_error(std::move(this->base()), std::current_exception());
+          }
+        }
+
+       public:
+        explicit __receiver(_Receiver __rcvr, _Shape __shape, _Fun __fun)
+          : receiver_adaptor<__receiver, _Receiver>((_Receiver&&) __rcvr)
+          , __shape_(__shape)
+          , __f_((_Fun&&) __fun)
+        {}
+      };
+
+    template <class _SenderId, integral _Shape, class _FunId>
+      struct __sender {
+        using _Sender = __t<_SenderId>;
+        using _Fun = __t<_FunId>;
+        template <receiver _Receiver>
+          using __receiver = __receiver<__x<remove_cvref_t<_Receiver>>, _Shape, _FunId>;
+
+        [[no_unique_address]] _Sender __sndr_;
+        [[no_unique_address]] _Shape __shape_;
+        [[no_unique_address]] _Fun __fun_;
+
+        template <class _Tag, class _Fun, class _Sender, class _Env>
+          using __with_error_invoke_t =
+            __if_c<
+              __v<__gather_sigs_t<
+                _Tag,
```

Comment:
`_Tag` can only be `set_value_t` here, so `__gather_sigs_t<_Tag,...` can be replaced with `__value_types_of_t<...`.